varname/utils.py

```diff
@@ -225,3 +230,75 @@ def debug_ignore_frame(
         msg = (f'{msg} [In {frameinfo.function!r} at '
                f'{frameinfo.filename}:{frameinfo.lineno}]')
     sys.stderr.write(f'[{__name__}] DEBUG: {msg}\n')
+
+def funcsig_from_ast_call(node: ast.Call,
+                          globs: dict,
+                          locs: dict) -> inspect.Signature:
+    """Get the function signature from an ast.Call node"""
+    expr = ast.Expression(node.func)
+    code = compile(expr, '<ast-call>', 'eval')
+    func = eval(code, globs, locs) # pylint: disable=eval-used
+    return inspect.signature(func)
+
+def argnode_source(source: Source, node: ast.AST, vars_only: bool) -> str:
+    """Get the source of an argument node"""
+    if vars_only:
+        if not isinstance(node, ast.Name):
+            raise NonVariableArgumentError(
+                f'Argument {ast.dump(node)} is not a variable.'
+            )
+        return node.id
+    return source.asttokens().get_text(node)
+
+@lru_cache()
+def get_argument_sources(node: ast.Call,
+                         vars_only: bool = True) -> MutableMapping[str, str]:
+    """Get the sources for argument from an ast.Call node
+
+    >>> def func(a, b, c, d=4):
+    >>>  ...
+    >>> x = y = z = 1
+    >>> func(y, x, c=z)
+    >>> # argument_sources = {'a': 'y', 'b', 'x', 'c': 'z'}
+    """
+    from .ignore import IgnoreList
+    frame = IgnoreList.create(ignore_lambda=False).get_frame(2)
+    # <Signature (a, b, c, d=4)>
+    signature = funcsig_from_ast_call(node, frame.f_globals, frame.f_locals)
```

Comment:
You can't put this under lru_cache, the function being called might be different for different calls of the same node.